cmd/object-handlers.go

```diff
@@ -534,13 +535,55 @@ func (api objectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
 		return
 	}
 
-	// Copy source object to destination, if source and destination
-	// object is same then only metadata is updated.
-	objInfo, err := objectAPI.CopyObject(ctx, srcBucket, srcObject, dstBucket, dstObject, srcInfo)
-	if err != nil {
-		pipeWriter.CloseWithError(err)
-		writeErrorResponse(w, toAPIErrorCode(err), r.URL)
-		return
+	var objInfo ObjectInfo
+
+	_, err = objectAPI.GetBucketInfo(ctx, dstBucket)
+	if err == toObjectErr(errVolumeNotFound, dstBucket) && !cpSrcDstSame {
+		if globalDNSConfig != nil {
+			if dstRecord, errEtcd := globalDNSConfig.Get(dstBucket); errEtcd == nil {
+				go func() {
+					if err = objectAPI.GetObject(ctx, srcBucket, srcObject, 0, srcInfo.Size, srcInfo.Writer, srcInfo.ETag); err != nil {
+						pipeWriter.CloseWithError(err)
+						writeErrorResponse(w, ErrInternalError, r.URL)
+						return
+					}
+					// Close writer explicitly to indicate data has been written
+					defer srcInfo.Writer.Close()
+				}()
+				// Forward request to appropriate instance
+				client, err := getRemoteInstanceClient(dstRecord)
+				if err != nil {
+					pipeWriter.CloseWithError(err)
+					writeErrorResponse(w, ErrInternalError, r.URL)
+					return
+				}
+				if _, err = client.PutObject(dstBucket, dstObject, srcInfo.Reader, srcInfo.Size, "", "", srcInfo.UserDefined); err != nil {
+					pipeWriter.CloseWithError(err)
+					writeErrorResponse(w, ErrInternalError, r.URL)
+					return
+				}
+				remoteObjInfo, err := client.StatObject(dstBucket, dstObject, miniogo.StatObjectOptions{})
+				if err != nil {
+					pipeWriter.CloseWithError(err)
+					writeErrorResponse(w, ErrInternalError, r.URL)
+					return
+				}
+				objInfo.ETag = remoteObjInfo.ETag
+				objInfo.ModTime = remoteObjInfo.LastModified
+			}
+		} else {
+			writeErrorResponse(w, ErrNoSuchBucket, r.URL)
+			return
+		}
+	} else {
```

Comment:
Another thing to add here is that we shouldn't do `GetObject()` and `PutObject()` style if both dest and source buckets are in the same location. Currently, this relies on only `cpSrcDstSame` this variable doesn't capture this situation. To make it a bit cleaner we should actually verify this on the buckets first under federated deployment. 

Basically 

 1. Verify if both source and bucket are in the same location, nothing needs to be done proceed with object layer copyObject API
 2. If they are different then proceed to use the combination of Get and Put. 
 3. If both buckets match in their names we don't need to do any DNS lookup let the call proceed as is in the previous situations. 

This should be perhaps done right at the beginning of the API and we need to also make sure we are handling encryption calls properly. If not we can add it in next iteration.  For all of this a separate function is better.

Reply:
I will add a separate function for this

Reply:
added a method, can you PTAL